clients/src/main/java/org/apache/kafka/server/authorizer/DefaultResourceTypeBasedAuthorization.java

```diff
@@ -0,0 +1,144 @@
+        Map<PatternType, Set<String>> allowPatterns = initializePatternMap();
+
+        boolean hasWildcardAllow = false;
+        KafkaPrincipal principal = createKafkaPrincipal(requestContext);
+        String hostAddr = requestContext.clientAddress().getHostAddress();
+
+        for (AclBinding binding : authorizer.acls(createAclBindingFilter(resourceType))) {
+            if (!isValidBinding(binding, principal, hostAddr, op)) {
+                continue;
+            }
+
+            if (binding.entry().permissionType() == AclPermissionType.DENY) {
+                processDenyAcl(binding, denyPatterns);
+                continue;
+            }
+
+            if (binding.entry().permissionType() == AclPermissionType.ALLOW) {
+                processAllowAcl(binding, allowPatterns);
+                if (binding.pattern().name().equals(ResourcePattern.WILDCARD_RESOURCE)) {
+                    hasWildcardAllow = true;
+                }
+            }
+        }
+
+        if (hasWildcardAllow) {
+            return AuthorizationResult.ALLOWED;
+        }
+
+        return evaluateAuthorizationResult(allowPatterns, denyPatterns);
+    }
+
+    private static boolean isSuperUserAuthorized(AuthorizableRequestContext requestContext, AclOperation op, ResourceType resourceType, Authorizer authorizer) {
+        return authorizer.authorize(requestContext, Collections.singletonList(new Action(
+                        op, new ResourcePattern(resourceType, "hardcode", PatternType.LITERAL),
+                        0, true, false)))
+                .getFirst() == AuthorizationResult.ALLOWED;
+    }
+
+    private static Map<PatternType, Set<String>> initializePatternMap() {
+        return new EnumMap<>(PatternType.class) {{
+            put(PatternType.LITERAL, new HashSet<>());
+            put(PatternType.PREFIXED, new HashSet<>());
+        }};
+    }
+
+    private static KafkaPrincipal createKafkaPrincipal(AuthorizableRequestContext requestContext) {
+        return new KafkaPrincipal(
+                requestContext.principal().getPrincipalType(),
+                requestContext.principal().getName());
+    }
+
+    private static AclBindingFilter createAclBindingFilter(ResourceType resourceType) {
+        ResourcePatternFilter resourcePatternFilter = new ResourcePatternFilter(resourceType, null, PatternType.ANY);
+        return new AclBindingFilter(resourcePatternFilter, AccessControlEntryFilter.ANY);
+    }
+
+    private static boolean isValidBinding(AclBinding binding, KafkaPrincipal principal, String hostAddr, AclOperation op) {
+        return (binding.entry().host().equals(hostAddr) || binding.entry().host().equals("*"))
+                && (SecurityUtils.parseKafkaPrincipal(binding.entry().principal()).equals(principal) || binding.entry().principal().equals("User:*"))
+                && (binding.entry().operation() == op || binding.entry().operation() == AclOperation.ALL);
```

Comment:
Since you did a good change, here I would also consider decrease cognitive load during read this part of code

Reply:
Thx, simplified in latest commit